nixos/systemd-boot: Add basic test

danielfullmer opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
I extracted this from a more complicated PR testing systemd-boot functionality: https://github.com/NixOS/nixpkgs/pull/84204

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

danielfullmer committed:
nixos/systemd-boot: Add basic test

flokli commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -189,9 +189,14 @@ let
           echo '(hd0) /dev/vda' > /boot/grub/device.map
 
-          # Install GRUB and generate the GRUB boot menu.
+          # Install bootloader
```
So this is not grub-specific anymore? I still see grub-specifics above…

danielfullmer replied:
Rearranged comments to make this more clear.

flokli replied:
Thanks, this makes much more sense now!

flokli commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -189,9 +189,14 @@ let
+          # Install bootloader
           touch /etc/NIXOS
           mkdir -p /nix/var/nix/profiles
+          ln -s ${config.system.build.toplevel} /nix/var/nix/profiles/system-1-link
+          ln -s /nix/var/nix/profiles/system-1-link /nix/var/nix/profiles/system
```
This is not done by the `switch-to-configuration` script?

danielfullmer replied:
Nope. Normally, this is done by `nixos-rebuild` which uses `nix-env -p /nix/var/nix/profile/system --set ...` before running `switch-to-configuration`.

flokli commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -189,9 +189,14 @@ let
+          ln -s ${config.system.build.toplevel} /nix/var/nix/profiles/system-1-link
+          ln -s /nix/var/nix/profiles/system-1-link /nix/var/nix/profiles/system
+          export NIXOS_INSTALL_BOOTLOADER=1
+          export SYSTEMD_RELAX_ESP_CHECKS=1
```
This needs a comment what we do here, and why.

danielfullmer replied:
Addressed in later comment

flokli commented on nixos/tests/all-tests.nix:
```diff
@@ -291,6 +291,7 @@ in
   syncthing-relay = handleTest ./syncthing-relay.nix {};
   systemd = handleTest ./systemd.nix {};
   systemd-analyze = handleTest ./systemd-analyze.nix {};
+  systemd-boot = handleTestOn ["x86_64-linux"] ./systemd-boot.nix {};
```
I don't think this is x86_64-linux specific. systemd-boot should work on aarch64, too.

danielfullmer replied:
Indeed, it would be nice to extend this to aarch64. Currently I think `qemu-vm.nix` only works with x86_64. We'd need to use `AAVMF` instead of `OVMF.fd`, among possibly other things. I think this should be in another PR, however.

flokli replied:
I think it does work - at least we build `nixosTests.installer.simpleUefiSystemdBoot` for aarch64, too: https://hydra.nixos.org/job/nixos/trunk-combined/nixos.tests.installer.simpleUefiSystemdBoot.aarch64-linux

danielfullmer replied:
So the installer test `simpleUefiSystemdBoot` actually has some special logic for aarch64 which does not currently have an analogue in `qemu-vm.nix`:
https://github.com/NixOS/nixpkgs/blob/694ac1b127ce34fb3bf5890232e1375b3911867c/nixos/tests/installer.nix#L72

It doesn't look very complicated, but didn't immediately work when I tried moving that logic into `qemu-vm.nix`.  I'll keep working on it, but I think it'd be better to put it in a new PR where we try to get `nixos-rebuild build-vm` with UEFI working on aarch64.

flokli replied:
Fine to touch this in a future PR :-)

flokli commented on nixos/tests/systemd-boot.nix:
```diff
@@ -0,0 +1,30 @@
+
+    # Ensure we actually booted using systemd-boot
+    machine.succeed(
+        "test -e /sys/firmware/efi/efivars/LoaderEntrySelected-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f"
```
Where is this magic number coming from?

danielfullmer replied:
It's a fixed UUID used by systemd-boot:
https://www.freedesktop.org/wiki/Software/systemd/BootLoaderInterface/

flokli replied:
Thanks for adding the comment!

flokli reviewed (commented):
Sorry, this fell off my radar. Some comments.

danielfullmer commented:
@floki Just added a new commit addressing the feedback.  I also got rid of `SYSTEMD_RELAX_ESP_CHECKS` by instead adding a link for `/dev/block/254:2` so systemd-boot can detect the ESP partition and properly set up the vars needed for EFI boot.  Previously, I think it was actually falling back to the BIOS boot. (It was also needed for the tests I've been working on for secure boot).

flokli commented:
@danielfullmer can you squash the two commits into one? Otherwise, this LGTM.

danielfullmer commented:
@flokli OK. Squashed, rebased against master, and tested building `nixosTests.systemd-boot`.

flokli commented:
Thanks!
